### PR TITLE
Fluid api

### PR DESCRIPTION
**Pull request by BoogieMonster1O1**
Taken from https://github.com/BoogieMonster1O1/nomcfluids

For information on how this works, read fluid-volume-api-debating in the fabricord

### CODE REVIEW

**Review by CoolMineman — CHANGES_REQUESTED** [review]
No test mod.
Not even an impl for Cauldrons.
Unclear how you even lookup FluidStorage from a world and pos.
There does not appear any way to revert a FluidTransaction which defeats the purpose of even having it.
FluidStorage has a questionable api where setStored directly mutates the underlying values and is the only way of inserting or extracting fluids.
The API provides multiple hard-coded checks (isValid and getMax[Input/Output]) the caller must check before mutating the underlying value instead of having an extract and insert and simply asking if the insertion or extraction can be done or not (through a simulation) or providing a rollback mechanism through a proper transaction system.
I fail to see the value in this API at all.

**Comment by hYdos** [comment]
my review is basicly summed up by trollzers.

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/impl/fluid/volume/ImmutableFluidSettings.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.awt.Color;
+import java.util.function.IntSupplier;
+import java.util.function.Supplier;
+
+import net.legacyfabric.fabric.api.fluid.volume.v1.settings.DensityUnits;
+import net.legacyfabric.fabric.api.fluid.volume.v1.settings.FluidSettings;
+import net.legacyfabric.fabric.api.fluid.volume.v1.settings.TemperatureUnits;
+
+public class ImmutableFluidSettings implements FluidSettings {
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/impl/fluid/volume/ImmutableFluidSettings.java:28]: no

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/util/Side.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+package net.legacyfabric.fabric.api.fluid.volume.v1.util;
+
+import java.util.Optional;
+
+import javax.annotation.Nonnull;
+
+import net.minecraft.util.math.Direction;
+
+public enum Side {
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/util/Side.java:26]: yes

File: fabric-fluid-volume-api-v1/src/test/java/net/legacyfabric/fabric/test/fluid/volume/FluidTransactionTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import net.legacyfabric.fabric.api.fluid.volume.v1.store.FluidStorage;
+import net.legacyfabric.fabric.api.fluid.volume.v1.store.base.SimpleFluidRate;
+import net.legacyfabric.fabric.api.fluid.volume.v1.util.Side;
+
+public class FluidTransactionTest {
+	@Test
+	public void testSingleFluidTransaction() {
+		Fraction max = new Fraction(1000);
+		Fraction twenty = new Fraction(20);
+		FluidType water = FluidType.getOrCreate(Blocks.WATER, FluidSettings.builder().luminance(0).color(Color.BLUE).build());
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/test/java/net/legacyfabric/fabric/test/fluid/volume/FluidTransactionTest.java:44]: no

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/FluidType.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+import net.minecraft.block.Block;
+import net.minecraft.block.FluidBlock;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.SimpleRegistry;
+
+import net.legacyfabric.fabric.api.fluid.volume.v1.settings.FluidSettings;
+
+public class FluidType {
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/FluidType.java:35]: no

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/FluidType.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	public static Optional<FluidType> get(Identifier id) {
+		return Optional.ofNullable(FLUID_TYPES.get(id));
+	}
+
+	public static FluidType getOrCreate(FluidBlock block, @Nullable FluidSettings settings) {
+		Identifier id = Block.REGISTRY.getIdentifier(block);
+		Preconditions.checkNotNull(id, "block was not registered");
+		return get(id).orElseGet(() -> new FluidType(id, block, Objects.requireNonNull(settings)));
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/FluidType.java:87]: yes

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/store/FluidTransaction.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.util.function.BooleanSupplier;
+
+import net.legacyfabric.fabric.api.fluid.volume.v1.fraction.Fraction;
+
+/**
+ * A wrapper around two {@link FluidStorageHandler} instances
+ * aimed to simplify transferring of fluid between two containers.
+ */
+public class FluidTransaction {
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/store/FluidTransaction.java:28]: transaction is a nice name

File: fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/store/FluidStorageHandler.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+			return Fraction.ZERO;
+		}
+
+		Fraction stored = this.storage.getStored(this.fluidType, this.side);
+		Fraction maxExtracted = amount.isGreaterThan(stored) ? amount : stored;
+		Fraction maxOutput = this.storage.getMaxOutput(this.fluidType, this.side);
+		Fraction extracted = maxExtracted.isGreaterThan(maxOutput) ? maxOutput : maxExtracted;
+
+		if (this.shouldPerform()) {
+			this.storage.setStored(this.fluidType, this.side, stored.withSubtraction(extracted));
```
[the remark this answers is not in the document]
Comment (BoogieMonster1O1) [reply fabric-fluid-volume-api-v1/src/main/java/net/legacyfabric/fabric/api/fluid/volume/v1/store/FluidStorageHandler.java:88]: if only the first is valid but the other isnt its empty

**Comment by BoogieMonster1O1** [comment]
wasting my time on something that isn't needed

**Pull request closed without merging**